mm2src/coins/rpc_command/get_estimated_fees.rs

```diff
@@ -0,0 +1,413 @@
+            };
+            // use returned bool (instead of run_state) to check if shutdown started to exit quickly
+            if shutdown_started.is_ok() && shutdown_started.unwrap() {
+                break;
+            }
+
+            let mut run_state = self.run_state.lock().await;
+            if let FeeEstimatorState::Stopping = *run_state {
+                *run_state = FeeEstimatorState::Stopped;
+                break;
+            }
+            drop(run_state);
+
+            let estimated = match estimated_res {
+                Ok(estimated) => estimated,
+                Err(_) => FeePerGasEstimated::default(), // TODO: if fee estimates could not be obtained should we clear values or use previous?
+            };
+            let mut estimated_fees = self.estimated_fees.lock().await;
+            *estimated_fees = estimated;
+            drop(estimated_fees);
+
+            let elapsed = common::now_float() - started;
+            debug!(
+                "{FEE_ESTIMATOR_NAME} getting estimated values processed in {} seconds",
+                elapsed
+            );
+
+            let wait_secs = FeeEstimatorContext::get_refresh_interval() - elapsed;
+            let wait_secs = if wait_secs < 0.0 { 0.0 } else { wait_secs };
+            let sleep_fut = Timer::sleep(wait_secs).fuse();
+            let stop_fut = self.listen_for_stop().fuse();
+            let shutdown_started = select! {
+                _ = Box::pin(sleep_fut) => Ok(false),
+                shutdown_started = Box::pin(stop_fut) => shutdown_started
+            };
+            if shutdown_started.is_ok() && shutdown_started.unwrap() {
+                break;
+            }
+        }
+        debug!("{FEE_ESTIMATOR_NAME} stopped");
+        Ok(())
+    }
+
+    async fn get_estimated_fees(ctx: MmArc, coin: &EthCoin) -> Result<FeePerGasEstimated, MmError<FeeEstimatorError>> {
+        Self::check_if_coin_connected(ctx.clone(), coin).await?;
+        let estimator_ctx = Self::from_ctx(ctx.clone())?;
+        let estimated_fees = estimator_ctx.estimated_fees.lock().await;
+        Ok(estimated_fees.clone())
+    }
+
+    async fn check_if_coin_connected(ctx: MmArc, coin: &EthCoin) -> Result<(), MmError<FeeEstimatorError>> {
+        let estimator_ctx = Self::from_ctx(ctx.clone())?;
+        let using_coins = estimator_ctx.using_coins.lock().await;
+        if using_coins.get(&coin.ticker().to_string()).is_none() {
+            return MmError::err(FeeEstimatorError::CoinNotConnected);
+        }
+        Ok(())
+    }
+}
+
+// Rpc request/response/result
+
+#[derive(Deserialize)]
+pub struct FeeEstimatorStartStopRequest {
+    coin: String,
+}
+
+#[derive(Serialize)]
+pub struct FeeEstimatorStartStopResponse {
+    result: String,
+}
+
+impl FeeEstimatorStartStopResponse {
+    #[allow(dead_code)]
+    pub fn get_result(&self) -> String { self.result.clone() }
```

Comment:
nit suggestion: I typically prefer returning the reference of the value and letting the caller decide whether a clone is needed or not. For example, if caller needs only the reference of this value, with this approach you will clone it regardless.

Reply:
changed to ref returned